# ETL_Pipeline_Project_06_25

This is my first project in which I am building an ETL pipeline in Python. My previous experience in building an ETL pipeline lies in SQL, however as we start to work with big data, utilising pandas and dataframes become essential in building highly optimised pipelines.

For this project I will improve a dataset from Kaggle (https://www.kaggle.com/datasets/ahmedmohamed2003/cafe-sales-dirty-data-for-cleaning-training).
<p>
This dataset contain 10,000 rows of data representing sales transactions in a cafe. The purpose of this project is to extract and clean the data and demonstrate data engineering ability and techniques.

## Extract

In [2]:
import pandas as pd
import math as m
import numpy as np
import matplotlib.pyplot as plt

In [3]:
# import the csv file.

cafe_sales_df = pd.read_csv('dirty_cafe_sales.csv')

In [4]:
cafe_sales_df.head(10)

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2            2.0         4.0   
1    TXN_4977031      Cake        4            3.0        12.0   
2    TXN_4271903    Cookie        4            1.0       ERROR   
3    TXN_7034554     Salad        2            5.0        10.0   
4    TXN_3160411    Coffee        2            2.0         4.0   
5    TXN_2602893  Smoothie        5            4.0        20.0   
6    TXN_4433211   UNKNOWN        3            3.0         9.0   
7    TXN_6699534  Sandwich        4            4.0        16.0   
8    TXN_4717867       NaN        5            3.0        15.0   
9    TXN_2064365  Sandwich        5            4.0        20.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
2     Credit Card  In-store       2023-07-19  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
5     Credit Card       NaN       2023-03-31  
6           ERROR  Takeaway       2023-10-06  
7            Cash   UNKNOWN       2023-10-28  
8             NaN  Takeaway       2023-07-28  
9             NaN  In-store       2023-12-31

In [5]:
cafe_sales_df.shape

(10000, 8)

In [6]:
cafe_sales_df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='object')

## Transform

In [7]:
cafe_sales_df

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
2       TXN_4271903    Cookie        4            1.0       ERROR   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9995    TXN_7672686    Coffee        2            2.0         4.0   
9996    TXN_9659401       NaN        3            NaN         3.0   
9997    TXN_5255387    Coffee        4            2.0         8.0   
9998    TXN_7695629    Cookie        3            NaN         3.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9995             NaN   UNKNOWN       2023-08-30  
9996  Digital Wallet       NaN       2023-06-02  
9997  Digital Wallet       NaN       2023-03-02  
9998  Digital Wallet       NaN       2023-12-02  
9999            Cash  In-store       2023-11-07  

[10000 rows x 8 columns]

In [8]:
cafe_sales_df.dropna(axis=0, inplace=True)
cafe_sales_df

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
2       TXN_4271903    Cookie        4            1.0       ERROR   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
...             ...       ...      ...            ...         ...   
9984    TXN_3142496  Smoothie  UNKNOWN            4.0         4.0   
9986    TXN_2858441  Sandwich        2            4.0         8.0   
9991    TXN_3897619  Sandwich        3            4.0        12.0   
9992    TXN_2739140  Smoothie        4            4.0        16.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
2        Credit Card  In-store       2023-07-19  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
...              ...       ...              ...  
9984            Cash  Takeaway       2023-07-27  
9986     Credit Card  In-store       2023-12-14  
9991            Cash  Takeaway       2023-02-24  
9992         UNKNOWN  In-store       2023-07-05  
9999            Cash  In-store       2023-11-07  

[4550 rows x 8 columns]

The new df has 4550 rows, I have removed each row which contains a missing values as this just adds noise to our dataset.

In [9]:
unique_item = cafe_sales_df['Item'].unique()
print(unique_item)
print('--------------------------------------------------------------------')
unique_payment = cafe_sales_df['Payment Method'].unique()
print(unique_payment)
print('--------------------------------------------------------------------')
unique_location = cafe_sales_df['Location'].unique()
print(unique_location)
print('--------------------------------------------------------------------')
#unique_date = cafe_sales_df['Transaction Date'].unique()
#print(unique_date)

['Coffee' 'Cake' 'Cookie' 'Salad' 'UNKNOWN' 'Sandwich' 'Juice' 'Smoothie'
 'ERROR' 'Tea']
--------------------------------------------------------------------
['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR']
--------------------------------------------------------------------
['Takeaway' 'In-store' 'UNKNOWN' 'ERROR']
--------------------------------------------------------------------


In the code above we see that our records contains values which from a business perspective is not useful to us. These rows have to be removed for us to build a table which we can analyse

Remove rows which contain error

In [10]:
filtered_cafe_sales_df = cafe_sales_df[ (cafe_sales_df['Location'] != 'ERROR') &
                                        (cafe_sales_df['Payment Method'] != 'ERROR') & 
                                        (cafe_sales_df['Item'] != 'ERROR') & 
                                        (cafe_sales_df['Total Spent'] != 'ERROR')  &
                                        (cafe_sales_df['Transaction Date'] != 'ERROR')  &
                                        (cafe_sales_df['Price Per Unit'] != 'ERROR')]

In [11]:
filtered_cafe_sales_df.head()

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0    TXN_1961373    Coffee        2            2.0         4.0   
1    TXN_4977031      Cake        4            3.0        12.0   
3    TXN_7034554     Salad        2            5.0        10.0   
4    TXN_3160411    Coffee        2            2.0         4.0   
7    TXN_6699534  Sandwich        4            4.0        16.0   

   Payment Method  Location Transaction Date  
0     Credit Card  Takeaway       2023-09-08  
1            Cash  In-store       2023-05-16  
3         UNKNOWN   UNKNOWN       2023-04-27  
4  Digital Wallet  In-store       2023-06-11  
7            Cash   UNKNOWN       2023-10-28

In [12]:
filtered_cafe_sales_df.shape

(3804, 8)

Conduct a quick check below to make sure we have removed the error transactions.

In [13]:
unique_item = filtered_cafe_sales_df['Item'].unique()
print(unique_item)
print('--------------------------------------------------------------------')
unique_payment = filtered_cafe_sales_df['Payment Method'].unique()
print(unique_payment)
print('--------------------------------------------------------------------')
unique_location = filtered_cafe_sales_df['Location'].unique()
print(unique_location)

['Coffee' 'Cake' 'Salad' 'Sandwich' 'Juice' 'Smoothie' 'Cookie' 'Tea'
 'UNKNOWN']
--------------------------------------------------------------------
['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet']
--------------------------------------------------------------------
['Takeaway' 'In-store' 'UNKNOWN']


## Load

Now that we cleaned our dataset, we will proceed to load this into a clean table ready to be used for analysis. Once we load the data into a clean table, we will export it as a csv file which can then be connected to by Power BI or Excel.

In [14]:
final_cafe_sales_df = filtered_cafe_sales_df
final_cafe_sales_df

Transaction ID      Item Quantity Price Per Unit Total Spent  \
0       TXN_1961373    Coffee        2            2.0         4.0   
1       TXN_4977031      Cake        4            3.0        12.0   
3       TXN_7034554     Salad        2            5.0        10.0   
4       TXN_3160411    Coffee        2            2.0         4.0   
7       TXN_6699534  Sandwich        4            4.0        16.0   
...             ...       ...      ...            ...         ...   
9984    TXN_3142496  Smoothie  UNKNOWN            4.0         4.0   
9986    TXN_2858441  Sandwich        2            4.0         8.0   
9991    TXN_3897619  Sandwich        3            4.0        12.0   
9992    TXN_2739140  Smoothie        4            4.0        16.0   
9999    TXN_6170729  Sandwich        3            4.0        12.0   

      Payment Method  Location Transaction Date  
0        Credit Card  Takeaway       2023-09-08  
1               Cash  In-store       2023-05-16  
3            UNKNOWN   UNKNOWN       2023-04-27  
4     Digital Wallet  In-store       2023-06-11  
7               Cash   UNKNOWN       2023-10-28  
...              ...       ...              ...  
9984            Cash  Takeaway       2023-07-27  
9986     Credit Card  In-store       2023-12-14  
9991            Cash  Takeaway       2023-02-24  
9992         UNKNOWN  In-store       2023-07-05  
9999            Cash  In-store       2023-11-07  

[3804 rows x 8 columns]

In [19]:
# export the final dataframe to a CSV.

filepath = r'C:\Users\cemil\Documents\Git Repo\ETL_Project_06_25\final_cafe_sales.csv'
final_cafe_sales_df.to_csv(filepath, sep=',')